### DISCUSSION TITLE
HUD takes time to display when large skin animations are used

### DISCUSSION THREAD

**Discussion by KevinMarkRoy** [ppy/osu#25282] [Q&A]
### Type

Performance

### Bug description

_OS Build_: Windows 10 Pro - 22H2 - 19045.3570
_GPU_: RTX 3060 (Driver - 537.13)
_CPU_: Ryzen 5 3600 (not overclocked, default settings)
_RAM_: 2x8 gigs 2133 Mhz

This issue occurs almost every time I start the first map after turning on the PC or restarting, it's rare to find an instance where it does not happen. If I restart osu, but don't shut down or restart my PC, the issue does not occur again. 

The delay in the video I have pasted is quite short but sometimes it can go upto 2-3 seconds. Whilst this delay is happening, inputs from the keyboard are not registered properly either. 

### Screenshots or videos

https://github.com/ppy/osu/assets/47815628/d804144e-8f46-4fa2-9e2b-62164abe0e71

### Version

2023.1026.0-lazer

### Logs

[performance.log](https://github.com/ppy/osu/files/13195620/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/13195621/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/13195622/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/13195623/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/13195624/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/13195625/runtime.log)
[updater.log](https://github.com/ppy/osu/files/13195626/updater.log)
[database.log](https://github.com/ppy/osu/files/13195627/database.log)
[input.log](https://github.com/ppy/osu/files/13195628/input.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/13195629/legacy-ipc.log)
[network.log](https://github.com/ppy/osu/files/13195631/network.log)

**Comment by bdach** [comment]
Please attach a video of this with frame graph shown (ctrl-f11 twice).

**Reply by KevinMarkRoy** [reply]
Let me know if the quality is too terrible, compressed it a lot to fit the 10MB upload limit.



https://github.com/ppy/osu/assets/47815628/6cbbc617-8d1a-4299-b487-9f0b518fa1c8



[performance-draw.log](https://github.com/ppy/osu/files/13196687/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/13196688/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/13196689/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/13196690/runtime.log)
[updater.log](https://github.com/ppy/osu/files/13196691/updater.log)
[database.log](https://github.com/ppy/osu/files/13196692/database.log)
[input.log](https://github.com/ppy/osu/files/13196693/input.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/13196694/legacy-ipc.log)
[network.log](https://github.com/ppy/osu/files/13196695/network.log)
[performance.log](https://github.com/ppy/osu/files/13196696/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/13196697/performance-audio.log)

**Comment by peppy** [comment]
First guess would be the animated health bar in the skin. Please attach/link your skin.

**Reply by KevinMarkRoy** [reply]
[Logia Skin (31-10-22) (Logia).zip](https://github.com/ppy/osu/files/13197471/Logia.Skin.31-10-22.Logia.zip)

**Reply by peppy** [reply]
Yeah, skin is going to be the issue. 1,216 frames of animations.

Immediate improvement may be to load each HUD element in parallel, rather than blocking the whole display. But that doesn't seem to work in practice (maybe not enough load threads?):

```diff
diff --git a/osu.Game/Screens/Play/HUDOverlay.cs b/osu.Game/Screens/Play/HUDOverlay.cs
index 128f8d5ffd..64175096a2 100644
--- a/osu.Game/Screens/Play/HUDOverlay.cs
+++ b/osu.Game/Screens/Play/HUDOverlay.cs
@@ -6,6 +6,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading;
 using JetBrains.Annotations;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
diff --git a/osu.Game/Skinning/LegacyHealthDisplay.cs b/osu.Game/Skinning/LegacyHealthDisplay.cs
index 845fc77394..647b72cf69 100644
--- a/osu.Game/Skinning/LegacyHealthDisplay.cs
+++ b/osu.Game/Skinning/LegacyHealthDisplay.cs
@@ -4,6 +4,7 @@
 #nullable disable
 
 using System;
+using System.Threading;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
@@ -36,6 +37,8 @@ private void load(ISkinSource source)
         {
             AutoSizeAxes = Axes.Both;
 
+            Thread.Sleep(5000);
+
             var skin = source.FindProvider(s => getTexture(s, "bg") != null);
 
             // the marker lookup to decide which display style must be performed on the source of the bg, which is the most common element.
diff --git a/osu.Game/Skinning/SkinComponentsContainer.cs b/osu.Game/Skinning/SkinComponentsContainer.cs
index adf0a288b4..b0e1dfdde5 100644
--- a/osu.Game/Skinning/SkinComponentsContainer.cs
+++ b/osu.Game/Skinning/SkinComponentsContainer.cs
@@ -66,20 +66,31 @@ public void Reload(Container? componentsContainer)
             components.Clear();
             ComponentsLoaded = false;
 
-            content = componentsContainer ?? new Container
+            AddInternal(content = new Container
             {
                 RelativeSizeAxes = Axes.Both
-            };
+            });
 
             cancellationSource?.Cancel();
             cancellationSource = null;
 
-            LoadComponentAsync(content, wrapper =>
+            if (componentsContainer == null)
+                return;
+
+            foreach (var component in componentsContainer)
             {
-                AddInternal(wrapper);
-                components.AddRange(wrapper.Children.OfType<ISerialisableDrawable>());
-                ComponentsLoaded = true;
-            }, (cancellationSource = new CancellationTokenSource()).Token);
+                LoadComponentAsync(componentsContainer, _ =>
+                {
+                    componentsContainer.Remove(component, false);
+                    content.Add(component);
+
+                    if (component is ISerialisableDrawable serialisableDrawable)
+                        components.Add(serialisableDrawable);
+
+                    if (componentsContainer.All(c => c.IsLoaded))
+                        ComponentsLoaded = true;
+                }, (cancellationSource = new CancellationTokenSource()).Token);
+            }
         }
 
         /// <inheritdoc cref="ISerialisableDrawableContainer"/>

```

We should probably revisit the load performance of such animations as it should be faster than the user is seeing. Personally I can't reproduce the slowdown with the same skin – the HUD loads immediately for me.

**Reply by KevinMarkRoy** [reply]
This issue only happens on the first map of the first osu session. If I just restart osu without restarting my PC, this does not happen to me.

**Reply by peppy** [reply]
Yep, things get cached after the first load.

**Reply by KevinMarkRoy** [reply]
I took out the healthbar from my skin using the layout editor but the issue still persists. Are all elements of the base skin being loaded regardless of it being placed in the layout?
